docs(misc): Add federate module recipe (#19642)
nrwl/nx: docs/shared/recipes/module-federation/federate-a-module.md

# Federate a Module

> Module Federation is a feature that allows to share code between applications at run-time. It is a way of doing micro-frontends, but it can also be used to share code between applications that are not micro-frontends.

Reviewer: ```suggestion
Module Federation is a concept that allows developers to share code between applications at run-time. It is a way of doing micro-frontends, but it can also be used to share code between applications that are not micro-frontends.
```

In the context of Module Federation, a _module_ can be thought as any piece of code, functionally independent, that can be reused in different applications. It can be a component, a service, a library, a utility, etc.

In order to share a module, it must be _federated_. This means that the module must be configured to be shared, and the application that wants to use it must be configured to use it.

**Nx** includes first class support for Module Federation for React and Angular applications. This means that you can federate modules in your workspace with a few simple commands.

{% callout type="info" title="Setup" %}

> With this recipe we assume that you have already created a workspace with at least one React or Angular module federation application.

Reviewer: ```suggestion
With this recipe we assume that you have already created a workspace with at least one React or Angular Module Federation host application.
```

> If you haven't, you can follow the [Create a Host recipe](https://nx.dev/recipes/module-federation/create-a-host).

Reviewer: ```suggestion
If you haven't, you can follow the [Create a Host Recipe](https://nx.dev/recipes/module-federation/create-a-host).
```

{% /callout %}

## Step 1: Create the module

We will create a module that exports a hello function.
Since we are using Nx, we will create a library for this module.

**Create a library**

```shell
nx generate @nx/js:library --name=hello --unitTestRunner=jest --projectNameAndRootFormat=as-provided
```

Update the `hello.ts` file with the following code:

```typescript {% fileName="hello/src/lib/hello.ts" %}
export default function hello(): string {
  return 'Hello from Nx';
}
```

Update `hello/` barrel file `index.ts` with the following code:

```typescript {% fileName="hello/src/index.ts" %}
export { default } from './lib/hello';
```

## Step 2: Federate the module

Now that we have created the module, we need to configure it to be federated.

{% tabs %}
{%tab label="React"%}

```shell
nx generate @nx/react:federate-module Hello --remote=greeting --path=hello/src/index.ts --projectNameAndRootFormat=as-provided
```

{% /tab %}
{%tab label="Angular"%}

```shell
nx generate @nx/angular:federate-module Hello --remote=greeting --path=hello/src/index.ts --projectNameAndRootFormat=as-provided
```

{% /tab %}
{% /tabs %}

{% callout type="note" title="Remote does not exist" %}

> If the remote provided does not exist (in this instance _greeting_), Nx will create it for you.

Reviewer: ```suggestion
If the remote provided does not exist (in this instance _greeting_), Nx will create it for you. Otherwise, Nx will add it to the `exposes` property in an existing remote.
```

Author: Are you sure we need this? 
We demonstrate that it gets added to the `exposes` property below.

Reviewer: I think for clarity it makes sense to have it. Just to eliminate any kind of confusion. I don't mind a little extra verbosity in docs when it comes to things like this. It should also help the AI assistant if users were to ask it

{% /callout %}

This command will:

> - Add a module entry to the `greeting` remote module federation config file.

Reviewer: ```suggestion
- Adds a module entry to the `greeting` remote module federation config file.
```

{% tabs %}

> {%tab label="typescript"%}

Reviewer: ```suggestion
{%tab label="TypeScript Config File"%}
```

```typescript {% fileName="greeting/module-federation.config.ts" %}
import { ModuleFederationConfig } from '@nx/webpack';

const config: ModuleFederationConfig = {
  name: 'greeting',
  exposes: {
    './Module': './src/remote-entry.ts',
    './Hello': 'hello/src/index.ts', // <-- this line was added,
  },
};
export default config;
```

{% /tab %}

> {%tab label="javascript"%}

Reviewer: ```suggestion
{%tab label="JavaScript Config File"%}
```

```javascript {% fileName="greeting/module-federation.config.js" %}
module.exports = {
  name: 'greeting',
  exposes: {
    './Module': './src/remote-entry.ts',
    './Hello': 'hello/src/index.ts', // <-- this line was added
  },
};
```

{% /tab %}
{% /tabs %}

> - Add a typescript path mapping to the `greeting/Hello` module in `tsconfig.json` file.

Reviewer: ```suggestion
- Adds a TypeScript path mapping to the `greeting/Hello` module in your root TSConfig file.
```

```json {% fileName="/tsconfig.base.json" %}
{
  "paths": {
    "greeting/Module": ["greeting/src/remote-entry.ts"],
    "greeting/Hello": ["hello/src/index.ts"] // <-- this line was added
  }
}
```

## Step 3: Use the module

Update the host application to use the federated module.

```tsx {% fileName="home/src/app/app.tsx" %}
import hello from 'greeting/Hello';

export function App() {
  return <div>{hello()}</div>;
}

export default App;
```

> Similarly, you would update the Angular application to use the federated module.

Reviewer: ```suggestion
If you are using Angular, you would update the application in a similar fashion to use the federated module.
```

{% /callout %}

This command will:

- Add a module entry to the `greeting` remote module federation config file.

{% tabs %}
{%tab label="typescript"%}

```typescript {% fileName="greeting/module-federation.config.ts" %}
import { ModuleFederationConfig } from '@nx/webpack';

const config: ModuleFederationConfig = {
  name: 'greeting',
  exposes: {
    './Module': './src/remote-entry.ts',
    './Hello': 'hello/src/index.ts', // <-- this line was added,
  },
};
export default config;
```

{% /tab %}
{%tab label="javascript"%}

```javascript {% fileName="greeting/module-federation.config.js" %}
module.exports = {
  name: 'greeting',
  exposes: {
    './Module': './src/remote-entry.ts',
    './Hello': 'hello/src/index.ts', // <-- this line was added
  },
};
```

{% /tab %}
{% /tabs %}

- Add a typescript path mapping to the `greeting/Hello` module in `tsconfig.json` file.

```json {% fileName="/tsconfig.base.json" %}
{
  "paths": {
    "greeting/Module": ["greeting/src/remote-entry.ts"],
    "greeting/Hello": ["hello/src/index.ts"] // <-- this line was added
  }
}
```

## Step 4: Use the module

> Update the greeting remote application to use the federated module.

Reviewer: I would use the federated module in the host application, rather than the same remote it’s being exposed from

Author: hmm okay sure I can update that

Reviewer: You might also have to include adding the new `remote` "greeting" to the `host` application's `module-federation.config` file

Author: I'll add a comment but this should've been done for you when you create the the host.
